File: BasicRxJavaSample/app/src/main/java/com/example/android/observability/demoTest/ui/ImitateUserViewModel.java
package com.example.android.observability.demoTest.ui;

import android.util.Log;

import androidx.lifecycle.ViewModel;

import com.example.android.observability.demoTest.ImitateUserDataSource;
import com.example.android.observability.demoTest.persisitence.ImitateUser;

import io.reactivex.Completable;
import io.reactivex.Flowable;

public class ImitateUserViewModel extends ViewModel {
    private ImitateUserDataSource mDataSource;
    private ImitateUser mImitateUser;

    public ImitateUserViewModel(ImitateUserDataSource dataSource) {
        mDataSource = dataSource;
    }

    public Flowable<String> getUserName() {
        return mDataSource.getUser().map(imitateUser -> {
            mImitateUser = imitateUser;
            return imitateUser.getUserName();
        });
    }

   public Completable setUserName(String userName) {
//        Log.i("ImitateUserActivity", "setUserName: " + userName);
        mImitateUser = mImitateUser == null ? new ImitateUser(userName) : new ImitateUser(mImitateUser.getId(), userName);
        return mDataSource.insertOrUpdateUser(mImitateUser);
    }
}

File: BasicRxJavaSampleKotlin/app/src/main/java/com/example/android/observability/imitate/ui/ImitateUserViewModel.kt
package com.example.android.observability.imitate.ui

import androidx.lifecycle.ViewModel
import com.example.android.observability.imitate.persistance.ImitateUser
import com.example.android.observability.imitate.persistance.ImitateUserDataSource
import io.reactivex.Completable
import io.reactivex.Flowable

class ImitateUserViewModel(imitateUserDataSource: ImitateUserDataSource) : ViewModel() {
    private var imitateUserDataSource: ImitateUserDataSource = imitateUserDataSource

    public fun setUserName(userName: String) :Completable{
        var imitateUser = ImitateUser("1",userName)
       return imitateUserDataSource.updateOrInsertUser(imitateUser)
    }

    public fun getUser(): Flowable<String> {
       return imitateUserDataSource.getUser().map {
           it.name
       }
    }

}